### ISSUE TITLE
Budgie Desktop View 0.99.3 segfault

### ISSUE THREAD

**Issue by leapfog** [getsolus/budgie-desktop-view#6]
- go to ~/Desktop
- create file **crash.desk** with following content
```
[Desktop Entry]
Version=1.0
Type=Link
Name=fileserver
Icon=folder-remote-smb
URL=smb://fileserver
```
- rename **crash.desk** to **crash.desktop**
- all desktop icons are gone

**Comment by NickJ2911** [comment]
Here is the contents of ~/Desktop

 image.jpeg
'url.desktop'
 folder
 file.pdf
 folder
'url.desktop'
 document2.pdf
 document.pdf
'url.desktop'
'url.desktop'
 plain text document
'url.desktop'
'url.desktop'
 folder
'folder'
'url.desktop'
'url.desktop'
"folder"
'file.pdf'
'folder'
 file.zip
 file.pdf
'url.desktop'
'url.desktop'
'url.desktop'
'url.desktop'
 folder
 folder
'url.desktop'
'url.desktop'
'url.desktop'
'Screenshot.png'
'Screenshot.png'
'Screenshot.png'
'Screenshot.png'
'url.desktop'
'url.desktop'
'url.desktop'
'url.desktop'
 Screenshot.png
'url.desktop'
'url.desktop'
'url.desktop'
'url.desktop'
'url.desktop'

I renamed them what they are and left the extension. All the the .desktop files are URLs that I dragged and dropped onto my desktop. I know that is kind of weird but its something that I like to do for quick access instead of bookmarking. I was reading that drag and drop wont be available until desktop-view 1.1. Is that something that could possibly be causing this? Also, If I move all of my desktop files to another location, such as documents could that solve the issue?

**Comment by JoshStrobl** [comment]
@NickJ2911 Yea it likely has to do with the URL. Interestingly enough, I created a shortcut in Google Chrome and it generated a valid desktop file which opens the URL in Google Chrome. Can I get a copy of one of your desktop files?

**Comment by JoshStrobl** [comment]
@leapfog I'm able to reproduce the segfault with that entry and will see about getting it fixed in a `1.0.1` release.

**Comment by JoshStrobl** [comment]
Update: The issue is the fact that those desktop files, while perfectly valid in their own right, are `Link` type Desktop files rather than `Application`, and so reasonably `DesktopAppInfo` can't handle them. I'm going to leverage try loading the files in as a `KeyFile` and just pulling the data I need out of them.

Discovered another issue that takes a slightly higher precedent, so this is going in 1.0.2 and going to push out a separate 1.0.1 that will at least prevent the segfault.

**Comment by JoshStrobl** [comment]
1.0.1 has been pushed to Solus stable, be sure to update and log out / log back in. Those non-Application-type items will currently not be rendered and will warn on being invalid, but you shouldn't get a segfault anymore. But no need to go through your desktop files and change everything, working on the KeyFile stuff now :)

**Comment by NickJ2911** [comment]
Thanks Josh! Update worked! Icons are showing up on desktop. Minus the .desktop URL web pages. I had a question about right clicking. In this new desktop-view you only have the option of open and open in terminal. Before you had many more options such as rename, delete, and move to trash. Is that something that will be available in future updates? Also, PDF files had a preview desktop icon of the actual file like the .jpg and .png files have now. Now the PDFs only have the pdf logo on the icons. The icons are also locked in place, but I presume that is what you are going to address in your drag and drop update. Thanks again!

**Comment by JoshStrobl** [comment]
> Before you had many more options such as rename, delete, and move to trash. Is that something that will be available in future updates?

This is because they were handled by Nautilus, itself being a file manager. Budgie Desktop View is not a file manager, per our README, its purposes are much more focused in scope. Deleting files can be recursive, which requires custom logic there, renames would involve custom handling, etc. All of these instead can be done through the file manager.

> Now the PDFs only have the pdf logo on the icons.

PDFs likely used tracker to generate them. We don't use tracker. JPGs, PNGs, SVGs are generated using Pixbuf which handles the conversion.

> The icons are also locked in place, but I presume that is what you are going to address in your drag and drop update. 

No. I'm not adding custom positioning. We're using a GtkFlowbox which has a specific sort order function, which puts items automatically in specific rows and columns. I don't want to complicate the implementation by moving it to a GtkGrid.

**Comment by JoshStrobl** [comment]
@NickJ2911 I'd like a copy of one of your desktop files with the URLs so I can validate it. Already validated @leapfog's example SMB one.

**Comment by NickJ2911** [comment]
@JoshStrobl Thank you for the detailed explanation. Very logical and makes sense. Attached is one of my .desktop URL files for you to validate. I compressed it into a .zip so I could add to this reply.
[Releases · GloriousEggroll\proton-ge-custom · GitHub.zip](https://github.com/getsolus/budgie-desktop-view/files/5640086/Releases.GloriousEggroll.proton-ge-custom.GitHub.zip)

**Comment by JoshStrobl** [comment]
Validated that my fix works. Thanks!

**Comment by JoshStrobl** [comment]
@leapfog @NickJ2911 I've uploaded an eopkg with e48fa06 as a patch to the Solus site (in our sources folder) if you'd like to validate it:

```
sudo eopkg install https://getsol.us/sources/budgie-desktop-view-1.0.1-13-1-x86_64.eopkg
```

After this, log out and log back in. Then try one of your previous desktop items. Thanks!

**Comment by leapfog** [comment]
When I run `xdg-open smb://remotehost/dir/` in my terminal I also get an `gio: smb://remotehost/dir/: The specified location is not mounted` error message.

Running `nautilus smb://remotehost/dir/` brings a password dialog, and after I connected anonymously, the Desktop link works as expected.

**Comment by JoshStrobl** [comment]
Yes, we don't perform the mounting, because that would require us to implement the password dialog and mount logic ourselves. Which is why we only show active mounts :)

But sounds like it is working otherwise, so going to consider this resolved.

**Comment by leapfog** [comment]
In fact I had the host without a directory in my original link. In that case xdg-open initially complains about not being mounted while nautilus shows that remote hosts available shares. When I run xdg-open after nautilus, that also shows the available shares. Same is with the desktop file: So it seems nautilus has to run against that remote host once, then it works.

If a directory (AKA share) is also specified in the desktop file, then it only works while that share is mounted. After unmounting it, the desktop file stops working.

BTW: Shouldn't the mounted shares be shown by budgie-desktop-view (even without having a desktop file)?

**Comment by JoshStrobl** [comment]
As a side note, I'd suggest making sure to install the latest 1.0.2 release. You're not going to get the update otherwise since you have a budgie-desktop-view installed that'll match the relnum (not to be confused with the version). Easiest way to do this is run (this is from the official unstable repo):

```
sudo eopkg install https://mirrors.rit.edu/solus/packages/unstable/b/budgie-desktop-view/budgie-desktop-view-1.0.2-13-1-x86_64.eopkg
```

> If a directory (AKA share) is also specified in the desktop file, then it only works while that share is mounted. After unmounting it, the desktop file stops working.

Yes as I explained, **we do not handle mounting, only showing active mounts.** This is not a bug.

> BTW: Shouldn't the mounted shares be shown by budgie-desktop-view (even without having a desktop file)?

Right now it's physical disk / removal media volumes, not network volumes. I can see about introducing support for it though, assuming it gets exposed via VolumeMonitor. I can test that personally, I have samba network shares locally :)

**Comment by leapfog** [comment]
Mayby I should simply rewrite my SMB-desktop files and turn them into Application ones that call nautilus with the given URL.

**Comment by leapfog** [comment]
I didn't want to nag, actually I saw your reply only after I send mine. (I've been interrupted writing it, so it took some time.)

As I said, I started thinking about rewriting the desktop files to run nautilus. Otherwise while I understand what happens, I'd have a hard time explaining my family, why that icon on the desktop sometimes works and sometimes doesn't.

Is there any reason why I shouldn't rewrite the desktop files (maybe because nautilus might be replaced soon)?

**Comment by leapfog** [comment]
I installed 1.0.2 and rebootet. And rewrote the desktop file to
```
[Desktop Entry]
Version=1.0
Name=SMB-Server
Type=Application
Icon=folder-remote
Exec=nautilus smb://smb-server/
```
Now everything works as expected.

**Comment by NickJ2911** [comment]
@JoshStrobl New patch works for me. Thanks again!

**Comment by JoshStrobl** [comment]
> Is there any reason why I shouldn't rewrite the desktop files (maybe because nautilus might be replaced soon)?

I don't expect us to replace Nautilus to be honest. I don't exactly have the time to commit to writing a file manager even if I wanted to immediate :D

But no there isn't really a reason not to rewrite them.

> I didn't want to nag, actually I saw your reply only after I send mine. (I've been interrupted writing it, so it took some time.)

No problem :)
